Add RRef support to ShardedTensor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59990 Remove PlacementSpec from ShardingSpecs.
* **#59776 Add RRef support to ShardedTensor.**

Overall design: https://github.com/pytorch/pytorch/issues/55207.

In this PR, I've added support to ShardedTensor such that it also creates RRefs
pointing to the remote shards if the RPC framework is initialized.

As a result, this provides more flexiblity for ShardedTensor such that users
can use collectives with local shards or use the RPC framework to interact with
remote shards.

Differential Revision: [D29020844](https://our.internmc.facebook.com/intern/diff/D29020844/)

pritamdamania committed:
Add RRef support to ShardedTensor.

Overall design: https://github.com/pytorch/pytorch/issues/55207.

In this PR, I've added support to ShardedTensor such that it also creates RRefs
pointing to the remote shards if the RPC framework is initialized.

As a result, this provides more flexiblity for ShardedTensor such that users
can use collectives with local shards or use the RPC framework to interact with
remote shards.

Differential Revision: [D29020844](https://our.internmc.facebook.com/intern/diff/D29020844/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -86,13 +113,14 @@ def __init__(
             raise ValueError(f'Global rank: {dist.get_rank()} not part of process group')
 
         self._local_shards: List[Shard] = []
+        self._remote_shards: List[rpc.RRef] = None
```
1) Why is the default value `None` instead of an empty list?
2) `rpc.RRef` -> `rpc.RRef[Shard]`?

pritamdamania87 replied:
1. It is mostly for the check in `remote_shards`. If it is None it means RPC was not initialized before ShardedTensor was created and we should throw an error instead of returning an empty list.
2. Will update this.

SciPioneer replied:
> It is mostly for the check in remote_shards. If it is None it means RPC was not initialized before ShardedTensor was created and we should throw an error instead of returning an empty list.

Thanks for the explanation! Saw the check in L358. I think it's still better to add a comment here to explain the subtle difference here, since readers may ignore the implementation details.

pritamdamania87 replied:
Added a separate field to track this instead for clarity

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -245,3 +346,18 @@ def size(self) -> torch.Size:
         """
         return torch.Size(self._dims)
+
+    def _register_remote_shards(self, remote_shards: List[rpc.RRef]):
```
Ditto

SciPioneer replied:
Another question is why is this a list instead of a dict keyed by name/rank? How shall we use it?

pritamdamania87 replied:
Good point, I'll change this to dict for easy of use.

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -113,6 +141,68 @@ def __init__(
+                fut = rpc.rpc_async(global_rank, _register_remote_shards, args=(all_tensor_ids[rank_to_name[global_rank]], rrefs))
+                futs.append(fut)
+
+        torch.futures.wait_all(futs)
```
Just want to confirm: the execution of each future here should be very fast right? I don't see any data movement, so adding each RRef should be pretty lightweight. Otherwise, we may need to consider deferring `wait_all` or only `wait` on individual remote shard on demand.

pritamdamania87 replied:
Yes, there is no data movement here. We are only sending RRefs to other processes so they know about all shards.

pritamdamania committed:
Update on "Add RRef support to ShardedTensor."


Overall design: https://github.com/pytorch/pytorch/issues/55207.

In this PR, I've added support to ShardedTensor such that it also creates RRefs
pointing to the remote shards if the RPC framework is initialized.

As a result, this provides more flexiblity for ShardedTensor such that users
can use collectives with local shards or use the RPC framework to interact with
remote shards.

Differential Revision: [D29020844](https://our.internmc.facebook.com/intern/diff/D29020844/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add RRef support to ShardedTensor."


Overall design: https://github.com/pytorch/pytorch/issues/55207.

In this PR, I've added support to ShardedTensor such that it also creates RRefs
pointing to the remote shards if the RPC framework is initialized.

As a result, this provides more flexiblity for ShardedTensor such that users
can use collectives with local shards or use the RPC framework to interact with
remote shards.

Differential Revision: [D29020844](https://our.internmc.facebook.com/intern/diff/D29020844/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -245,3 +346,18 @@ def size(self) -> torch.Size:
+    def _register_remote_shards(self, remote_shards: List[rpc.RRef]):
+        self._remote_shards.extend(remote_shards)
+
+    def remote_shards(self) -> List[rpc.RRef]:
```
Nit: add `@property` for the ease of access?

pritamdamania committed:
Update on "Add RRef support to ShardedTensor."


Overall design: https://github.com/pytorch/pytorch/issues/55207.

In this PR, I've added support to ShardedTensor such that it also creates RRefs
pointing to the remote shards if the RPC framework is initialized.

As a result, this provides more flexiblity for ShardedTensor such that users
can use collectives with local shards or use the RPC framework to interact with
remote shards.

Differential Revision: [D29020844](https://our.internmc.facebook.com/intern/diff/D29020844/)

[ghstack-poisoned]
